examples/graph/mpnn-molecular-graphs.py

```diff
@@ -0,0 +1,723 @@
+            #     tf.shape(atom_features_aggregated)[0]
+            # )
+            # if pad_len > 0:
+            #     atom_features_aggregated = tf.pad(
+            #         atom_features_aggregated, [(0, pad_len), (0, 0)]
+            #     )
+
+            # Update aggregated atom_features via a step of GRU
+            atom_features_updated, _ = self.update_step(
+                atom_features_aggregated, atom_features_updated
+            )
+
+        return atom_features_updated
+
+
+"""
+### Readout
+
+When the message passing procedure ends, the k-step-aggregated node states are to be
+reduced to a graph-level embedding. The easy alternative of performing this reduction is
+to use an average pooling layer. However, following the original paper, a set-to-set
+layer will be implemented. The set-to-set (*SetGather*) layer is a sequence-to-sequence layer
+for sets, for which the order does not matter. I.e., the resulting graph-level embedding
+is invariant to the order of the elements of the input sequence. The implementation is
+based on the paper
+[Order Matters: Sequence to sequence for sets](https://arxiv.org/abs/1511.06391) and
+[DeepChem's SetGather](https://deepchem.readthedocs.io/en/latest/api_reference/layers.html).
+"""
+
+
+class NoInputLSTMCell(keras.layers.Layer):
+    """Custom LSTM Cell that takes no input"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def build(self, input_shape):
+        memory_state_dim = input_shape[0][-1]
+        carry_state_dim = input_shape[1][-1]
+        self.recurrent_kernel = self.add_weight(
+            shape=(carry_state_dim, memory_state_dim * 4),
+            trainable=True,
+            initializer="glorot_uniform",
+        )
+        self.bias = self.add_weight(
+            shape=(memory_state_dim * 4), trainable=True, initializer="zeros",
+        )
+
+    def call(self, inputs):
+
+        # Unpack states
+        memory_state, carry_state = inputs
+
+        # Perform linear transformation on carry_state
+        z = tf.matmul(carry_state, self.recurrent_kernel) + self.bias
+
+        # Split transformed carry_state into four units (gates/states)
+        update_gate, forget_gate, memory_state_candidate, output_gate = tf.split(
+            z, num_or_size_splits=4, axis=1
+        )
+
+        # Apply non-linearity to all units
+        update_gate = tf.nn.sigmoid(update_gate)
+        forget_gate = tf.nn.sigmoid(forget_gate)
+        output_gate = tf.nn.sigmoid(output_gate)
+        memory_state_candidate = tf.nn.tanh(memory_state_candidate)
+
+        # Forget and update memory state
+        memory_state = forget_gate * memory_state + update_gate * memory_state_candidate
+
+        # Update carry state
+        carry_state = output_gate * tf.nn.tanh(memory_state)
+
+        # Return (updated) memory state and carry state
+        return memory_state, carry_state
+
+
+class SetGather(keras.layers.Layer):
```

Comment:
This is a very surprising architecture choice. The best tool for doing sequence-to-sequence over sets is to use a Transformer (without positional encoding). Transformer is the canonical set-processing building block.

Reply:
Yes indeed, I followed the [paper](https://arxiv.org/abs/1704.01212). Again, thanks a lot for the feedback, I'll look into it!